### ISSUE TITLE
Run simulations with different content placement algorithms

### ISSUE THREAD

**Issue by ICarvalho** [icarus-sim/icarus#22]
Hi Lorenzo,

I'd like to run experiments, but using different content placement algorithms in cache nodes. I'll describe what I'm trying to do:

Obs: I'm assuming that each content request is composed of a list of contents and not a single one. for example: req = [1, 3, 6]. Moreover, each request may be variable-length or fixed-length.

1 - Collect a data-set containing content requests used in a sample experiment. 
2 - Split the data-set into training and test. In the training one, I intend to find out patterns in such requests in order to perform a content placement based on them. Once the patterns are found, my proposal, in an off-line manner, places these content on cache nodes. So, my network.py file would be like this:
``` python
 policy_name = cache_policy['name']
        policy_args = {k: v for k, v in cache_policy.items() if k != 'name'}
        # The actual cache objects storing the content
        self.cache = {node: CACHE_POLICY[policy_name](cache_size[node], **policy_args)
                          for node in `cache_size}`
       for node in self.cache:
            contents = [
                           [1,2,3],
                           [4,5],
                           [5,6,7,8]
                             ]
             for content in contents:
                    self.cache[node].put_contents(random.choice(contents))
```
3 - At the test phase, the content requests are performed, but now having already placed the contents found in the training phase and compare the results with the other available proposals. 

I'm having some problems during simulation because when engine.py gets the NetworModel from network.py file to be used on  the simulation, all the policies to be evaluated will use the same content allocation (in this case, the one I obtained in the training phase and described above). What I'd like to have is that my proposal would use its own NetworkModel, with its own  content allocation scheme, and the other proposal would use the baseline one, provided on Icarus. 

I tried to design two different NetworkModels on network.py to be used by engine.py, but didn't work. Follow is a sample of my code:

obs: I replaced strategy by policy, but it is not the cache policy itself.

``` python
def exec_experiment(topology, workload, netconf, policy, cache_policy, collectors):

    model_baseline = NetworkModelBaseLine(topology, cache_policy, **netconf)
    model_proposal = NetworkModelProposal(topology, cache_policy, **netconf)
    view_baseline = NetworkViewBaseLine(model_baseline)
    view_proposal = NetworkViewProposal(model_proposal)
    controller_baseline = NetworkController(model_baseline)
    controller_proposal = NetworkController(model_proposal)


    collectors_inst_baseline = [DATA_COLLECTOR[name](view_baseline, **params)
                       for name, params in collectors.items()]
    collector_baseline = CollectorProxy(view_baseline, collectors_inst_baseline)
    controller_baseline.attach_collector(collector_baseline)

    policy_name_baseline = policy['name']
    policy_args_baseline = {k: v for k, v in policy_baseline.items() if k != 'name'  and policy_name_baseline == 'SAMPLE_POLICY'}
    policy_inst_baseline = POLICY[policy_name_baseline](view_baseline, controller_baseline, **policy_args_baseline)

    collectors_inst_proposal = [DATA_COLLECTOR[name](view_proposal, **params)
                                for name, params in collectors.items()]
    collector_proposal = CollectorProxy(view_proposal, collectors_inst_proposal)
    controller_proposal.attach_collector(collector_proposal)

    policy_name_proposal = policy_proposal['name']
    policy_args_proposal = {k: v for k, v in policy_proposal.items() if k != 'name' and policy_name_proposal== 'PROPOSAL'}
    policy_inst_proposal = POLICY[policy_name_proposal](view_proposal, controller_proposal, **policy_args_proposal)

    for time, event in workload:
         policy_inst_baseline.process_event(time, **event)
         policy_inst_proposal.process_event(time, **event)
    return collector_baseline.results(), collector_proposal.results()

``` 
Could you please help me how to make each policy uses its corresponding NetworkModel and run with the correct scenario?

Best Regards, 

Igor.

**Comment by lorenzosaino** [comment]
Hi Igor, could you please edit your comment so that the code is included into code blocks with correct indentation? It's hard to follow the code. Reference: https://docs.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks

**Comment by ICarvalho** [comment]
> Hi Igor, could you please edit your comment so that the code is included into code blocks with correct indentation? It's hard to follow the code. Reference: https://docs.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks

Hi, Lorenzo. 

Really sorry about that. Now, it's already fixed. 

Regards, 

Igor.

**Comment by lorenzosaino** [comment]
Hi Igor, I thought about this. There isn't a very clean way to do it with Icarus. I think the solution requiring the least amount of work to implement would be to run two experiments per scenario: you use the first experiment as training to collect data and the second experiment for testing. To do so, I think all you'll need to do is:
* Implement a data collector to get the data you need in the training experiment (if the ones you have already are not enough)
* Implement a content placement strategy for the test experiment that reads the result of the offline optimization and places content accordingly

**Comment by lorenzosaino** [comment]
I am closing the issue for now. Feel free to reopen it if you still have further issues.
